File: src/main/java/com/github/kornilova_l/flamegraph/plugin/server/converters/file_to_call_traces/flamegraph/TreesSetImpl.java
package com.github.kornilova_l.flamegraph.plugin.server.converters.file_to_call_traces.flamegraph;

import com.github.kornilova_l.flamegraph.plugin.server.trees.Filter;
import com.github.kornilova_l.flamegraph.plugin.server.trees.TreeManager;
import com.github.kornilova_l.flamegraph.plugin.server.trees.TreesSet;
import com.github.kornilova_l.flamegraph.proto.TreeProtos;
import com.github.kornilova_l.flamegraph.proto.TreesPreviewProtos;
import com.github.kornilova_l.flamegraph.proto.TreesProtos;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class TreesSetImpl extends TreesSet {
    public TreesSetImpl(@NotNull TreeProtos.Tree callTraces) {
        this.setCallTraces(callTraces);
    }

    @Override
    public TreesPreviewProtos.TreesPreview getTreesPreview(@Nullable Filter filter) {
        throw new UnsupportedOperationException("Call tree is not supported for .jfr");
    }

    @Nullable
    @Override
    public TreeProtos.Tree getTree(@NotNull TreeManager.TreeType treeType, Filter filter) {
        switch (treeType) {
            case BACK_TRACES:
                return getBackTracesMaybeFiltered(filter);
            case CALL_TRACES:
                return getCallTracesMaybeFiltered(filter);
            default:
                throw new IllegalArgumentException("Tree type " + treeType + " is unsupported");
        }
    }

    @Nullable
    @Override
    public TreesProtos.Trees getCallTree(@Nullable Filter filter) {
        throw new UnsupportedOperationException("Call tree is not supported for .jfr");
    }

    @Override
    public TreesProtos.Trees getCallTree(@Nullable Filter filter, @Nullable List<Integer> threadsIds) {
        throw new UnsupportedOperationException("Call tree is not supported for .jfr");
    }
}

File: src/main/java/com/github/korniloval/flameviewer/trees/TreesSetImpl.kt
package com.github.korniloval.flameviewer.trees

import com.github.korniloval.flameviewer.trees.util.TreePreviewBuilder
import com.github.korniloval.flameviewer.trees.util.accumulative.trees.call.traces.CallTreeToCallTracesConverter
import com.github.kornilova_l.flamegraph.proto.TreeProtos
import com.github.kornilova_l.flamegraph.proto.TreesPreviewProtos
import com.github.kornilova_l.flamegraph.proto.TreesProtos

class TreesSetImpl : TreesSet {
    constructor(callTraces: TreeProtos.Tree) {
        this.callTraces = callTraces
    }

    constructor(callTree: TreesProtos.Trees) {
        /* call traces will be computed when requested */
        this.callTree = callTree
    }

    override fun getTreesPreview(filter: Filter?): TreesPreviewProtos.TreesPreview? {
        val callTree = getCallTree(filter) ?: return null
        return TreePreviewBuilder(callTree).treesPreview
    }

    override fun getTree(treeType: TreeManager.TreeType, filter: Filter?): TreeProtos.Tree? {
        if (callTraces == null) {
            callTraces = CallTreeToCallTracesConverter(callTree!!).tree
        }
        return when (treeType) {
            TreeManager.TreeType.BACK_TRACES -> getBackTracesMaybeFiltered(filter)
            TreeManager.TreeType.CALL_TRACES -> getCallTracesMaybeFiltered(filter)
        }
    }

    override fun getCallTree(filter: Filter?): TreesProtos.Trees? {
        if (callTree == null) {
            return null
        }
        if (filter == null) {
            return callTree
        }
        val filteredTrees = TreesProtos.Trees.newBuilder()
        for (tree in callTree!!.treesList) {
            val filteredTree = filterTree(tree, filter, true)
            if (filteredTree != null) {
                filteredTrees.addTrees(filteredTree)
            }
        }
        return if (filteredTrees.treesCount == 0) {
            null
        } else filteredTrees.build()
    }

    override fun getCallTree(filter: Filter?, threadsIds: List<Int>?): TreesProtos.Trees? {
        val trees = getCallTree(filter) ?: return null
        if (threadsIds == null) {
            return trees
        }
        val wantedTrees = TreesProtos.Trees.newBuilder()
        for (threadsId in threadsIds) {
            wantedTrees.addTrees(trees.getTrees(threadsId))
        }
        return wantedTrees.build()
    }
}
